### PR TITLE
VT v3 Relationships support

### PR DESCRIPTION
**Pull request by orhovy**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/35621

## Description
Added new params to the integration: ip_relationships, url_relationships, domain_relationships, file_relationships
Added create relationships support.

## Screenshots
image

image

https://user-images.githubusercontent.com/58938354/118118719-f85aed00-b3f5-11eb-9809-ace7321dc3bb.mov

https://user-images.githubusercontent.com/58938354/117676327-1df9b380-b1b6-11eb-9cf3-3882ba7d4a97.mov

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [x] 6.2.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Documentation

### CODE REVIEW

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.yml [hunk 1]
```diff
@@ -100,6 +100,90 @@ configuration:
...
+  - '* pcap children'
+  - '* pcap parents'
+  - pe resource children
+  - pe resource parents
+  - '* similar files'
+  required: false
+  type: 16
+  defaultvalue: '* carbonblack children,* carbonblack parents,* compressed parents,contacted domains,contacted ips,contacted urls,dropped files,
+  * email attachments,* email parents,* embedded domains,* embedded ips,* embedded urls,execution parents,* itw domains,* itw ips,
+  * overlay children,* overlay parents,* pcap children,* pcap parents,pe resource children,pe resource parents,* similar files'
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.yml:186]: please use this opportunity to move insecure and proxy to the bottom.

File: Packs/VirusTotal/ReleaseNotes/2_1_0.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### VirusTotal (API v3)
+- Added support for relationships
```
Comment (bakatzir) [inline Packs/VirusTotal/ReleaseNotes/2_1_0.md:4]:
```suggestion
- Added support for relationships.
```

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3_test.py [hunk 3]
```diff
@@ -154,3 +154,29 @@ def test_get_working_id(self):
     def test_get_working_id_no_entry(self):
         with pytest.raises(DemistoException):
             assert get_working_id('1451', '')
+
+
+def test_create_relationships():
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3_test.py:159]: let's also add a test when no relationships/some are configured in the params, and only the relevant had been created.

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py [hunk 4]
```diff
@@ -94,6 +162,8 @@ def get_ip_comments(self, ip: str, limit: int) -> dict:
         See Also:
             https://developers.virustotal.com/v3.0/reference#ip-comments-get
         """
+        demisto.info('in get_ip_comments')
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:165]: leftovers?

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py [hunk 5]
```diff
@@ -1041,12 +1141,16 @@ def build_domain_output(
...
     data = raw_response.get('data', {})
     attributes = data.get('attributes', {})
+    relationships_response = data.get('relationships', {})
     whois: defaultdict = get_whois(attributes.get('whois', ''))
     score = score_calculator.domain_score(domain, raw_response)
     if score != Common.DBotScore.BAD and client.is_premium:
         score = score_calculator.analyze_premium_domain_score(client, domain, score)
     logs = score_calculator.get_logs()
     demisto.debug(logs)
+    relationships_list = create_relationships(entity_a=domain, entity_a_type='Domain',
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:1151]: let's use the CSP constants: FeedIndicatorType.Domain

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py [hunk 6]
```diff
@@ -1933,17 +2057,23 @@ def main(params: dict, args: dict, command: str):
     handle_proxy()
     client = Client(params)
     score_calculator = ScoreCalculator(params)
+
+    ip_relationships = (','.join(argToList(params.get('ip_relationships')))).replace('* ', '').replace(" ", "_")
+    url_relationships = (','.join(argToList(params.get('url_relationships')))).replace('* ', '').replace(" ", "_")
+    domain_relationships = (','.join(argToList(params.get('domain_relationships')))).replace('* ', '').replace(" ", "_")
+    file_relationships = (','.join(argToList(params.get('file_relationships')))).replace('* ', '').replace(" ", "_")
+    demisto.info(domain_relationships)
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:2065]: leftovers?

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py [hunk 7]
```diff
@@ -14,6 +14,74 @@
 COMMAND_PREFIX = "vt"
 INTEGRATION_ENTRY_CONTEXT = "VirusTotal"
 
+INDICATOR_TYPE = {
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:17]: is this needed? can you achieve the same functionality  with `FeedIndicatorType`?

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py [hunk 8]
```diff
@@ -35,44 +103,44 @@ def __init__(self, params: dict):
...
-    def ip(self, ip: str) -> dict:
+    def ip(self, ip: str, relationships: str = '') -> dict:
         """
         See Also:
             https://developers.virustotal.com/v3.0/reference#ip-info
         """
         return self._http_request(
             'GET',
-            f'ip_addresses/{ip}'
+            f'ip_addresses/{ip}?relationships={relationships}'
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:113]: relevant to all, if no `relationships` configured, will this call fail? let's add UT for that.

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.yml [hunk 9]
```diff
@@ -100,6 +100,90 @@ configuration:
   defaultvalue: '3'
   type: 0
   required: false
+- additionalinfo: 'Select the list of relationships which will be brought from the api. Some of the relationships are signed with * key which indicates that they are
```
Comment (bakatzir) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.yml:103]: for all of this params. they will only be relevant for 6.2 TIM customers, correct? we need to think of a way to reflect that in the docs/additionalinfo.

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py (hunk 8 above)
Comment (orhovy) [reply Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:113]: Are you asking regarding the api call? regarding its failure?
If so, I've checked it does not fail, if no relationships are set then the api does not return them.

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py (hunk 7 above)
Comment (orhovy) [reply Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:17]: In my opinion no, as FeedIndicatorType is only an object that contains the common types.
As we get the type from the api we need to convert the string to our convention of it which is done by this dict.
Do you think there is a better way?

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py [hunk 10]
```diff
@@ -979,6 +1048,36 @@ def analyze_premium_ip_score(self, client: Client, ip: str, base_score: int) ->
...
+    relationships_list: List[EntityRelationship] = []
+    for relationship_type, relationship_type_raw in relationships_response.items():
+
+        relationships_data = relationship_type_raw.get('data', [])
+        if isinstance(relationships_data, dict):
+            relationships_data = [relationships_data]
+
+        for relation in relationships_data:
+            name = RELATIONSHIP_TYPE.get(entity_a_type.lower(), {}).get(relationship_type)
+            entity_b = relation.get('id', '')
```
Comment (jochman) [inline Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:1071]: Why using the default `''`? just use `[]` and raise exception if not exists.
Do not keep running on the function if it does not exists. It's a bad practice that leads to a lot of bugs that are hard to catch. use [] and print the exception to the debug log.

Comment (orhovy) [reply Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3.py:1071]: We do want to keep running in the function if it doesnt exist, as the next one might exist.
If one entity_b doesnt exist it doesnt mean we want to raise any errors, just continue and create the other relationships .

File: Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3_test.py [hunk 11]
```diff
@@ -154,3 +154,29 @@ def test_get_working_id(self):
...
+
+    When:
+    - create relationships function.
+
+    Then:
+    - Validate that the relationships were created as expected.
+    """
+    expected_name = ['communicates-with', 'communicates-with', 'related-to', 'related-to']
+
+    with open("./TestData/ip.json", encoding='utf-8') as f:
```
[the remark this answers is not in the document]
Comment (orhovy) [reply Packs/VirusTotal/Integrations/VirusTotalV3/VirusTotalV3_test.py:172]: why is that ?

**Review by yuvalbenshalom — APPROVED** [review]
good work

**Review by bakatzir — APPROVED** [review]

**Comment by orhovy** [comment]
@ShirleyDenkberg Can you help with the additional info and the readme that was added ?

**Comment by ShirleyDenkberg** [comment]
@orhovy I reviewed the changes as you requested.  I was not able to upload my changes in the VirusTotal/Integration/Readme, so I sent the file to you via Slack. Please upload the contents of that file.

**Pull request merged**